lib/Parse/ParseExpr.cpp

```diff
@@ -2744,6 +2744,79 @@ void Parser::ParseBlockId(SourceLocation CaretLoc) {
   Actions.ActOnBlockArguments(CaretLoc, DeclaratorInfo, getCurScope());
 }
 
+IdentifierInfo *Parser::getBoundsExpressionStart() {
+  if (Tok.getKind() == tok::identifier) {
+    IdentifierInfo *Ident = Tok.getIdentifierInfo();
+    if (Ident == Ident_bounds || Ident == Ident_byte_count || Ident == Ident_count) {
+      return Ident;
+    }
+  }
+  return nullptr;
+}
+
+ExprResult Parser::ParseBoundsExpression() {
+  IdentifierInfo *Ident = getBoundsExpressionStart();
+  if (!Ident) {
+    Diag(Tok, diag::err_expected_bounds_expr);
+    return ExprError();
```

Comment:
I don't think there is a standard, but there is a convention.  Generally you should issue an error or diagnostic message before returning an ExprError().   Because this is compiler output, we don't have to worry about the message being lost.  In this method, most of the error outputs are buried in the helper methods: ExpectAndConsume, for example, issues an error message if the expected token does not appear next in the input.